### PR TITLE
WI #2650 Implement refactorings testing

### LINKED ISSUE #2650: Tooling: new tests for refactorings
LSR testing is cumbersome and refactoring needs more tests as it is complex to create edits on COBOL source.
The goal of this issue is to design and implement a new kind of tests specifically for our COBOL refactorings.

## Ideas:
- A refactoring test is made of 3 parts:
  - A target source code on which the refactoring must be applied
  - The description of a command (name + args)
  - The expected source code after refactoring has been applied

**Workflow of a test:**
- Parse test file, that means identify all 3 parts
  - Load and parse the target source code
  - Gather info to launch the command on the previously built AST
  - Apply modifications on source code, here we'll have to convert `TextEdit`s into `RangeUpdate`s
  - Read modified source code as text
  - Compare expected source code text with actual source code text

## Technical considerations:
- The method `ParserUtils.ParseCobolString` can be used to parse a chunk of text, however it must be extended to allow caller to specify the `ColumnsLayout` (and maybe other options if need be)
- Before testing refactoring independently from the server and the protocol, we need to create an abstraction for refactorings server side. In our current implementation commands directly compute edits and are tied to the `TypeCobolServer` object in order to retrieve the `CompilationUnit` being worked on and send request to client containing edits
  - A `Refactoring` may be seen as a specialization of a `Command` operating on a list of documents identified by their `Uri`s and depending on a `RefactoringProcessor` to produce the `TextEdits`.
  - In this model, the `RefactoringProcessor` class and its inheritors will be the elements being tested and therefore will have to be independent from the server.

### PR DESCRIPTION
**Pull request by fm-117**
Fixes #2650

### CODE REVIEW

**Review by efr15 — CHANGES_REQUESTED** [review]

File: TypeCobol.LanguageServer.Test/RefactoringTest.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                builder.AppendLine("      * 03: This line won't change");
+                builder.AppendLine("      * 04: This line will have some text deleted");
+                builder.AppendLine("      * 05: This line won't change");
+                builder.AppendLine("      * 06: This line will be modified");
+                builder.AppendLine("      * 07: This line won't change");
+                builder.AppendLine("      * 08: This line will have some text inserted");
+                builder.AppendLine("      * 09: This line won't change");
+                builder.AppendLine("      * 10: This line will be partially deleted");
+                builder.AppendLine("      * 11: This line will be deleted");
+                builder.AppendLine("      * 12: This line will be partially deleted");
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.LanguageServer.Test/RefactoringTest.cs:38]: It _seems_ but it isn't, the resulting line is `* 10: This line will be partially deleted` which is made of the beginning of line 10 and end of line 12. I'll update the test to show it more clearly.

File: TypeCobol.LanguageServer.Test/RefactoringTests/AdjustFillers/SimpleTest.cbl [hunk 2]
```diff
@@ -0,0 +1,48 @@
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. TCOMFL06.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 target-zone PIC X(200).
+       01 redef1 REDEFINES target-zone.
+          05 var1 PIC X(100).
+      * Ajuster ici la taille (20 -> 100)
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.LanguageServer.Test/RefactoringTests/AdjustFillers/SimpleTest.cbl:8]: Ok

File: TypeCobol.LanguageServer.Test/RefactoringTests/RefactoringProcessorTest.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            result.AppendLine($"refactoring.label={label}");
+            result.AppendLine("refactoring.source=");
+
+            // Apply text edits on original source code: we simulate here what happens client-side
+            int changeIndex = 0;
+            var nextChange = NextChange();
+            for (int lineIndex = 0; lineIndex < _target.CobolTextLines.Count; lineIndex++)
+            {
+                string lineText = _target.CobolTextLines[lineIndex].Text;
+                if (nextChange != null && lineIndex + 1 == nextChange.range.start.line)
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.LanguageServer.Test/RefactoringTests/RefactoringProcessorTest.cs:156]: Yes

File: TypeCobol.LanguageServer.Test/RefactoringTests/RefactoringProcessorTest.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+            // Parse original source code
+            var options = new TypeCobolOptions();
+            var format = DocumentFormat.RDZReferenceFormat;
+            bool isCopy = !testData.OriginalSource.TrimStart().StartsWith("IDENTIFICATION", StringComparison.OrdinalIgnoreCase); // Simple but should be enough, does not support copys starting with IDENTIFICATION...
+            var target = ParserUtils.ParseCobolString(testData.OriginalSource, isCopy, options, format);
+
+            // Identify processor to test
+            var refactoringProcessor = _RefactoringProcessors[testData.ProcessorType];
+
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.LanguageServer.Test/RefactoringTests/RefactoringProcessorTest.cs:50]: An exception would be thrown if the processor instance is not found.

**Review by efr15 — APPROVED** [review]

**Pull request merged**